tower-http/src/trace.rs

```diff
@@ -0,0 +1,348 @@
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let start = Instant::now();
+
+        let span = self.event_emitter.make_span(&req);
+
+        let future = {
+            let _guard = span.enter();
+            self.event_emitter.on_request_received(&req);
+            self.inner.call(req)
+        };
+
+        ResponseFuture {
+            future,
+            span,
+            start,
+            latency_unit: self.latency_unit,
+            classifier: self.classifier.clone(),
+            event_emitter: self.event_emitter.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// The [`Future`] produced by [`Trace`] services.
+///
+/// [`Future`]: std::future::Future
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, Classifier, EventEmitter, ReqBody> {
+    #[pin]
+    future: F,
+    span: Span,
+    start: Instant,
+    latency_unit: LatencyUnit,
+    classifier: Classifier,
+    event_emitter: EventEmitter,
+    _marker: PhantomData<fn() -> ReqBody>,
+}
+
+impl<F, ReqBody, ResBody, E, Classifier, EventEmitter> Future
+    for ResponseFuture<F, Classifier, EventEmitter, ReqBody>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    Classifier: ClassifyResponse<ResBody, E>,
+    EventEmitter:
+        EmitTracingEvents<ReqBody, OkClass = Classifier::OkClass, ErrClass = Classifier::ErrClass>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+
+        let _guard = this.span.enter();
+
+        let result = ready!(this.future.poll(cx));
+        let latency = this.start.elapsed();
+
+        let classification = this.classifier.classify_request_result(&result);
+        this.event_emitter
+            .on_classified_result(classification, latency, *this.latency_unit);
+
+        Poll::Ready(result)
+    }
+}
+
+/// Trait for emitting events and creating spans.
+///
+/// Designed to work with some implementation of [`ClassifyResponse`].
+pub trait EmitTracingEvents<ReqBody> {
+    /// The type used to classify successful responses.
+    ///
+    /// This mirrors [`ClassifyResponse::OkClass`].
+    type OkClass;
+
+    /// The type used to classify failed responses.
+    ///
+    /// This mirrors [`ClassifyResponse::ErrClass`].
+    type ErrClass;
+
+    /// Make a new [`tracing::Span`] from a request.
+    ///
+    /// [`tracing::Span`]: https://docs.rs/tracing/latest/tracing/struct.Span.html
+    fn make_span(&self, req: &Request<ReqBody>) -> Span;
+
+    /// Emit an event when a new request has been received.
+    fn on_request_received(&self, req: &Request<ReqBody>);
+
+    /// Emit an event when a request has been processed and the result has been classified.
+    fn on_classified_result(
+        &self,
+        result: ResponseClassification<Self::OkClass, Self::ErrClass>,
+        latency: Duration,
+        latency_unit: LatencyUnit,
+    );
```

Comment:
not totally sure how i feel about putting all these callbacks on one trait. it seems to me like it might be better to take a separate `Fn` for the span, request start, and response classification callbacks. that way, a user could override one of them while using the defaults for the others?

Reply:
Yeah I did consider that as well. However then we end up with like 4-5 type params on `Trace` which is quite a lot. Adding default implementations in the trait itself is hard because then we cannot pass configuration elsewhere than as direct arguments.